aten/src/ATen/native/cudnn/Conv.cpp

```diff
@@ -434,32 +452,38 @@ size_t getMaxWorkspaceSize(
 }
 
 template<typename perf_t>
-perf_t getBestAlgorithm(perf_t *perfResults, bool deterministic, int n_algo) {
-  if (deterministic) {
+perf_t getBestAlgorithm(perf_t *perfResults, ConvolutionArgs& args, int n_algo) {
+  int best_algo_idx;
+  if (args.params.deterministic) {
     // iterate over perf results of all algorithms and find the best deterministic algo
     for (int i = 0; i < n_algo; i++) {
       // TODO: Shouldn't all returned results be successful?
       // Double check documentation for cudnnFindConvolutionForwardAlgorithmEx
       if (perfResults[i].status == CUDNN_STATUS_SUCCESS &&
           perfResults[i].determinism == CUDNN_DETERMINISTIC) {
-        return perfResults[i];
+        best_algo_idx = i;
       }
     }
     AT_ERROR("no deterministic convolution algorithms available in CuDNN");
   } else {
-    return perfResults[0];
+    best_algo_idx = 0;
   }
+  // update convDesc mathType since cudnn 7.4+ now requires both algo + mathType to figure out
+  // whether to use Tensor core kernels or not
+  // See Note [behavior of cudnnFind and cudnnGet]
+  AT_CUDNN_CHECK(cudnnSetConvolutionMathType(args.cdesc.mut_desc(), perfResults[best_algo_idx].mathType));
```

Comment:
This won't work. This will set correct math type only the first time, when you are actually running find. All the subsequent times cdesc will be created with the default math type (default for fp32, tensor cores for fp16), and it won't ever be changed.

Reply:
Gotcha, I see what you are saying. I was trying to address @ezyang 's request on removing the following code blocks and this can either stay like this or I could make the setter take `perf_t` and do this there once.
```
if (dataType == CUDNN_DATA_HALF) {
    algoPerf->mathType = CUDNN_TENSOR_OP_MATH;
} else {
    algoPerf->mathType = CUDNN_DEFAULT_MATH;
}
```